src/battle_util.c

```diff
@@ -7835,25 +7852,32 @@ bool32 DoBattlersShareType(u32 battler1, u32 battler2)
 bool32 CanBattlerGetOrLoseItem(u8 battlerId, u16 itemId)
 {
     u16 species = gBattleMons[battlerId].species;
-
-    if (IS_ITEM_MAIL(itemId))
-        return FALSE;
-    else if (itemId == ITEM_ENIGMA_BERRY)
+    u16 holdEffect = ItemId_GetHoldEffect(itemId);
+    
+    // Mail can be stolen now
+    if (itemId == ITEM_ENIGMA_BERRY)
         return FALSE;
-    else if (species == SPECIES_KYOGRE && itemId == ITEM_BLUE_ORB)
+    else if (GET_BASE_SPECIES_ID(species) == SPECIES_KYOGRE && itemId == ITEM_BLUE_ORB) // includes primal
         return FALSE;
-    else if (species == SPECIES_GROUDON && itemId == ITEM_RED_ORB)
+    else if (GET_BASE_SPECIES_ID(species) == SPECIES_GROUDON && itemId == ITEM_RED_ORB) // includes primal
         return FALSE;
     // Mega stone cannot be lost if pokemon can mega evolve with it or is already mega evolved.
-    else if (ItemId_GetHoldEffect(itemId) == HOLD_EFFECT_MEGA_STONE
-             && ((GetMegaEvolutionSpecies(species, itemId) != SPECIES_NONE) || gBattleStruct->mega.evolvedPartyIds[GetBattlerSide(battlerId)] & gBitTable[gBattlerPartyIndexes[battlerId]]))
+    else if (holdEffect == HOLD_EFFECT_MEGA_STONE
+      && ((GetMegaEvolutionSpecies(species, itemId) != SPECIES_NONE)
+      || gBattleStruct->mega.evolvedPartyIds[GetBattlerSide(battlerId)] & gBitTable[gBattlerPartyIndexes[battlerId]]))
```

Comment:
```suggestion
    else if (GetMegaEvolutionSpecies(GET_BASE_SPECIES_ID(species), itemId) != SPECIES_NONE 
             && holdEffect == HOLD_EFFECT_MEGA_STONE)
```
to fix not being implemented:
```
- Pickpocket can't steal Mega Stone if attacker Pokemon is Mega Evolution of species that could have evolved with that Mega Stone
- Pickpocket can't steal Mega Stone if Pickpocketer Pokemon is Mega Evolution of species that could have evolved with that Mega Stone
```
For example, Mega Charizard X that is holding Charizardite X or Y cannot have it taken, but if holding Diancite can have it stolen from it. Another example: Mega Charizard X with Pickpocket cannot steal Charizardite X or Y from contact attacker, but can steal Diancite.